File: packages/rust/src/hashers/lsh.rs
//! SimHash LSH implementation (current default algorithm).

use crate::hasher::Hasher;
use crate::lsh::simhash_lsh_multi;
use crate::types::{LshConfig, LshFamily};
use crate::SigError;

/// SimHash LSH implementation
///
/// This is the current default algorithm using deterministic random hyperplanes
/// with SplitMix64 for reproducible signature generation.
pub struct SimHashLsh;

impl Hasher for SimHashLsh {
    fn name(&self) -> &str {
        "lsh"
    }

    fn compute(&self, embedding: &[f32], config: &LshConfig) -> Result<Vec<LshFamily>, SigError> {
        Ok(simhash_lsh_multi(embedding, config))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simhash_lsh_name() {
        let hasher = SimHashLsh;
        assert_eq!(hasher.name(), "lsh");
    }

    #[test]
    fn test_simhash_lsh_compute() {
        let hasher = SimHashLsh;
        let embedding = vec![0.1, 0.2, 0.3, 0.4];
        let config = LshConfig::default();

        let result = hasher.compute(&embedding, &config).unwrap();
        assert_eq!(result.len(), config.families);
        assert_eq!(result[0].bits, config.bits);
        assert_eq!(result[0].bands.len(), config.bands);
    }

    #[test]
    fn test_simhash_lsh_deterministic() {
        let hasher = SimHashLsh;
        let embedding = vec![0.1, 0.2, 0.3, 0.4];
        let config = LshConfig::default();

        let result1 = hasher.compute(&embedding, &config).unwrap();
        let result2 = hasher.compute(&embedding, &config).unwrap();

        assert_eq!(result1[0].signature, result2[0].signature);
    }
}

File: packages/typescript/src/hashers/lsh.ts
/**
 * SimHash LSH implementation (current default algorithm).
 * 
 * Uses deterministic random hyperplane LSH with SplitMix64-based
 * hyperplane generation. This is the canonical LSH implementation
 * used for V0 and V1 signatures.
 */

import type { Hasher } from '../hasher';
import { simhashLshMulti, type LSHFamily } from '../lsh';
import type { LshConfig } from '../types';

/**
 * SimHash LSH implementation.
 * 
 * @example
 * ```typescript
 * import { SimHashLsh } from '@0din/prompt-toolkit/hashers';
 * import { normalizeVector } from '@0din/prompt-toolkit';
 * 
 * const hasher = new SimHashLsh();
 * const vector = normalizeVector([1.0, 2.0, 3.0]);
 * const config = { families: 3, bits: 256, bands: 16 };
 * const families = hasher.compute(vector, config);
 * console.log(hasher.name()); // 'lsh'
 * ```
 */
export class SimHashLsh implements Hasher {
  /**
   * Return algorithm name.
   * @returns 'lsh'
   */
  name(): string {
    return 'lsh';
  }

  /**
   * Compute LSH signatures from normalized embedding.
   * 
   * Delegates to the simhashLshMulti function with the provided
   * configuration parameters.
   * 
   * @param embedding - L2-normalized embedding vector
   * @param config - LSH configuration parameters
   * @returns List of LSH families, one per family index
   */
  compute(embedding: number[], config: LshConfig): LSHFamily[] {
    return simhashLshMulti(embedding, config);
  }
}
